### DISCUSSION TITLE
Modify 'as' cast to allow for null-forgiveness immediately

### DISCUSSION THREAD

**Discussion by dylanrafael05** [dotnet/csharplang#5773] [Language Ideas]
This is a small request (although there may be issues with the internal implementation which make this impossible), but I have found myself writing the following very often and the parentheses seem extraneous.

```cs
var a = (x as string)!;
```

My proposal is to allow for the bang operator to be placed directly after the type in an 'as' cast.
```cs
var a = x as string!;
```

No further meaning is attached to this, it is simply a terser form. 
If '!' is to be used in the future for a sort of type-annotation, the '!' could be placed after 'as':
```cs
var a = x as! string;
```

Although this may be too much of a change and might over-encourage programmers from using null-forgiveness.

**Comment by CyrusNajmabadi** [comment]
Why are you writing this bit in the first place:

`var a = (x as string);` ?

`as` means: cast to this type, but if you cannot, give me back null.

So by using `as` you are saying that you really expect to have to deal with either a string or null . So using `!` after that is pretty fishy.  Can you clarify why you're using `as` if you don' want to have to deal with the potential null?

**Reply by dylanrafael05** [reply]
I mostly use 'as' as a substitute for c-style casting, mostly for sake of readability, as well as to reduce extra parenthesis.

**Reply by CyrusNajmabadi** [reply]
not sure how to help you here then :)  you're using the specific operator that exists precisely for the purpose of producing a value or null :)   

If you don't want to deal with null, then that's what the normal cast is for :)

**Reply by dylanrafael05** [reply]
That's fair enough. Hopefully in the future C# as a language can start to move away from c-style operations (case: break, (type)value) and towards better alternatives (case => {}, value as! type) so that it can take on a more modern flavor at its most basic levels.

Maybe, in this proposal, it would be more useful to let 'x as! string' be a synonym for '(string)x', and allow for things like 'x as! int'. This would make casting seem more consistent overall, although would create a greater learning curve.

**Reply by CyrusNajmabadi** [reply]
It might be better to continue the discussion over at: 

https://github.com/dotnet/csharplang/discussions/5447

**Comment by Rekkonnect** [comment]
I support this idea for the same reason why I prefer doing `A is 4` over `A == 4`. Both do the exact same thing and my personal style is to use the `is` keyword over the `==` operator.
